File: Sources/Buckett/ViewModels/AppState.swift
import AppKit
import Combine
import Foundation
import SwiftUI

enum SidebarSelection: Hashable {
    case dashboard
    case bucket(String)
}

@MainActor
final class AppState: ObservableObject {
    static let shared = AppState()

    let accountStore = AccountStore()
    let transfers = TransferManager()
    let updates = UpdateChecker()

    @Published var selectedAccountID: UUID? {
        didSet {
            UserDefaults.standard.set(selectedAccountID?.uuidString, forKey: "selectedAccountID")
        }
    }
    @Published var sidebarSelection: SidebarSelection? = .dashboard
    @Published var showOnboarding = false
    @Published var buckets: [Bucket] = []
    @Published var bucketsLoading = false
    @Published var bucketsError: String?
    /// Bucket lists for every account (not just the selected one), so the
    /// menu bar drop menu can offer buckets across accounts.
    @Published var accountBuckets: [UUID: [Bucket]] = [:]
    /// Bucket analytics keyed by "accountUUID|bucket" so every account's
    /// buckets can carry stats simultaneously.
    @Published var stats: [String: BucketStats] = [:]
    @Published var analyzing: Set<String> = []

    static func statsKey(_ accountID: UUID, _ bucket: String) -> String {
        accountID.uuidString + "|" + bucket
    }

    func stats(accountID: UUID, bucket: String) -> BucketStats? {
        stats[Self.statsKey(accountID, bucket)]
    }

    func isAnalyzing(accountID: UUID, bucket: String) -> Bool {
        analyzing.contains(Self.statsKey(accountID, bucket))
    }

    private var clientCache: [UUID: S3Client] = [:]
    private var cancellables: Set<AnyCancellable> = []
    private var statsRefreshTasks: [String: Task<Void, Never>] = [:]

    init() {
        // Re-publish nested store changes so views observing AppState refresh
        // when the account list changes.
        accountStore.objectWillChange
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &cancellables)

        // Refresh a bucket's stats shortly after uploads land in it.
        NotificationCenter.default.addObserver(
            forName: .buckettTransferCompleted, object: nil, queue: .main
        ) { [weak self] notification in
            guard let bucket = notification.userInfo?["bucket"] as? String else { return }
            Task { @MainActor [weak self] in
                self?.scheduleStatsRefresh(bucket: bucket)
            }
        }

        if let saved = UserDefaults.standard.string(forKey: "selectedAccountID"),
           let uuid = UUID(uuidString: saved),
           accountStore.accounts.contains(where: { $0.id == uuid }) {
            selectedAccountID = uuid
        } else {
            selectedAccountID = accountStore.accounts.first?.id
        }
    }

    var selectedAccount: Account? {
        accountStore.accounts.first { $0.id == selectedAccountID }
    }

    func client(for account: Account) -> S3Client? {
        if let cached = clientCache[account.id] { return cached }
        guard let client = accountStore.client(for: account) else { return nil }
        clientCache[account.id] = client
        return client
    }

    var currentClient: S3Client? {
        selectedAccount.flatMap { client(for: $0) }
    }

    // MARK: - Account lifecycle (invalidates cached clients)

    func saveAccount(_ account: Account, secretKey: String?) {
        accountStore.upsert(account, secretKey: secretKey)
        clientCache[account.id] = nil
        if selectedAccountID == nil {
            selectedAccountID = account.id
        }
        if selectedAccountID == account.id {
            Task { await loadBuckets() }
        }
    }

    func deleteAccount(_ account: Account) {
        accountStore.remove(account)
        clientCache[account.id] = nil
        if selectedAccountID == account.id {
            selectedAccountID = accountStore.accounts.first?.id
            buckets = []
            stats = [:]
            sidebarSelection = .dashboard
            Task { await loadBuckets() }
        }
    }

    func selectAccount(_ id: UUID?) {
        guard id != selectedAccountID else { return }
        selectedAccountID = id
        buckets = []
        stats = [:]
        bucketsError = nil
        sidebarSelection = .dashboard
        Task { await loadBuckets() }
    }

    // MARK: - Buckets

    func loadBuckets() async {
        guard let account = selectedAccount else {
            buckets = []
            return
        }
        guard let client = client(for: account) else {
            bucketsError = "No credentials found for “\(account.name)”. Add the secret key in Settings."
            buckets = []
            return
        }
        bucketsLoading = true
        bucketsError = nil
        do {
            buckets = try await client.listBuckets()
            accountBuckets[account.id] = buckets
            autoAnalyzeAll()
        } catch {
            bucketsError = error.localizedDescription
            buckets = []
        }
        bucketsLoading = false
        loadOtherAccountBuckets()
    }

    /// Background-refreshes bucket lists for accounts other than the selected
    /// one (menu bar drop menu, All Accounts sidebar/overview), then kicks off
    /// their stale-stats analysis.
    private func loadOtherAccountBuckets() {
        for account in accountStore.accounts where account.id != selectedAccountID {
            guard let client = client(for: account) else { continue }
            Task { [weak self] in
                guard let list = try? await client.listBuckets() else { return }
                self?.accountBuckets[account.id] = list
                self?.autoAnalyze(account: account)
            }
        }
    }

    func bucketList(for accountID: UUID) -> [Bucket] {
        if accountID == selectedAccountID, !buckets.isEmpty {
            return buckets
        }
        return accountBuckets[accountID] ?? []
    }

    func isValidDropTarget(_ target: MenuDropTarget) -> Bool {
        bucketList(for: target.accountID).contains { $0.name == target.bucket }
    }

    func createBucket(named name: String) async throws {
        guard let client = currentClient else { return }
        try await client.createBucket(name)
        await loadBuckets()
    }

    /// Deletes a bucket, optionally emptying it first. Emptying purges ALL
    /// object versions and delete markers (Backblaze B2 keeps hidden versions
    /// that make a bucket look empty while still blocking deletion) and aborts
    /// unfinished multipart uploads. Cleans up aliases, stats, and menu bar
    /// drop-target references afterwards.
    func deleteBucket(named name: String, in account: Account, emptyFirst: Bool) async throws {
        guard let client = client(for: account) else {
            throw S3Error(status: 0, code: "NoCredentials", message: "Missing account credentials")
        }
        if emptyFirst {
            let versions = try await client.listAllObjectVersions(bucket: name)
            if !versions.isEmpty {
                try await client.deleteObjectVersions(bucket: name, versions: versions)
            }
            let uploads = (try? await client.listMultipartUploads(bucket: name)) ?? []
            for upload in uploads {
                try? await client.abortMultipartUpload(
                    bucket: name, key: upload.key, uploadID: upload.uploadID
                )
            }
        }
        try await client.deleteBucket(name)

        stats.removeValue(forKey: Self.statsKey(account.id, name))
        BucketAliases.shared.setAlias(nil, accountID: account.id, bucket: name)
        let encoded = MenuDropTarget(accountID: account.id, bucket: name).encoded
        var shortlist = MenuBarController.dropShortlist()
        if shortlist.contains(encoded) {
            shortlist.removeAll { $0 == encoded }
            UserDefaults.standard.set(shortlist, forKey: MenuBarController.dropBucketsKey)
        }
        accountBuckets[account.id]?.removeAll { $0.name == name }
        if account.id == selectedAccountID {
            if sidebarSelection == .bucket(name) {
                sidebarSelection = .dashboard
            }
            await loadBuckets()
        }
    }

    // MARK: - Main window

    /// Captured from the SwiftUI environment so AppKit contexts (Dock reopen,
    /// menu bar "Open Buckett") can recreate the window after it was closed.
    var openWindowAction: OpenWindowAction?

    func openMainWindow() {
        NSApp.activate(ignoringOtherApps: true)
        if let visible = NSApp.windows.first(where: { $0.isVisible && $0.canBecomeKey }) {
            visible.makeKeyAndOrderFront(nil)
        } else {
            openWindowAction?(id: "main")
        }
    }

    // MARK: - Menu bar drops

    /// Where icon-drops go by default: the first checked drop-menu bucket (any
    /// account), else the selected bucket, else the account's first bucket.
    func menuBarAutoTarget() -> MenuDropTarget? {
        if let first = MenuBarController.dropShortlist()
            .compactMap({ MenuDropTarget(encoded: $0) })
            .first(where: { isValidDropTarget($0) }) {
            return first
        }
        guard let selectedID = selectedAccountID else { return nil }
        if case .bucket(let current) = sidebarSelection {
            return MenuDropTarget(accountID: selectedID, bucket: current)
        }
        if let first = buckets.first {
            return MenuDropTarget(accountID: selectedID, bucket: first.name)
        }
        return nil
    }

    /// Queues uploads dropped on the menu bar icon (or one of the hover panel's
    /// bucket rows, via `target` — which may belong to any account). Returns the
    /// resolved target for the animation, or nil (with feedback) when there is
    /// nowhere to upload.
    @discardableResult
    func handleMenuBarDrop(urls: [URL], to target: MenuDropTarget? = nil) -> MenuDropTarget? {
        guard let resolved = target ?? menuBarAutoTarget() else {
            openMainWindow()
            ToastCenter.shared.show(
                "No bucket available",
                detail: "Add an account and create a bucket, then drop files again.",
                style: .error
            )
            return nil
        }
        guard let account = accountStore.accounts.first(where: { $0.id == resolved.accountID }),
              let client = client(for: account) else {
            openMainWindow()
            ToastCenter.shared.show(
                "Missing credentials",
                detail: "Re-enter the secret key for this account in Settings.",
                style: .error
            )
            return nil
        }

        let bucket = resolved.bucket
        let transfers = self.transfers
        Task {
            let expanded = await Task.detached(priority: .userInitiated) {
                BrowserModel.expandForUpload(urls: urls)
            }.value
            for (fileURL, suffix) in expanded {
                transfers.enqueueUpload(
                    fileURL: fileURL, bucket: bucket, key: suffix, client: client
                )
            }
            Notifier.shared.post(
                .dropStarted,
                title: "Upload started",
                body: "\(expanded.count) file\(expanded.count == 1 ? "" : "s") → \(bucket)"
            )
        }
        return resolved
    }

    // MARK: - Analytics

    /// Convenience for the selected account.
    func analyze(bucket: String) {
        guard let account = selectedAccount else { return }
        analyze(account: account, bucket: bucket)
    }

    func analyze(account: Account, bucket: String) {
        Task { await analyzeNow(account: account, bucket: bucket) }
    }

    func analyzeNow(account: Account, bucket: String) async {
        let key = Self.statsKey(account.id, bucket)
        guard let client = client(for: account), !analyzing.contains(key) else { return }
        analyzing.insert(key)
        defer { analyzing.remove(key) }
        do {
            let objects = try await client.listAllObjects(bucket: bucket)
            stats[key] = Self.computeStats(bucket: bucket, objects: objects)
        } catch {
            NSLog("Buckett: analyze failed for \(bucket): \(error.localizedDescription)")
        }
    }

    /// Analyzes the account's buckets that have no stats yet or whose stats
    /// are older than 15 minutes. Sequential so accounts aren't hammered.
    func autoAnalyze(account: Account) {
        let staleBefore = Date().addingTimeInterval(-15 * 60)
        let names = bucketList(for: account.id).map(\.name)
        Task {
            for name in names {
                if let existing = stats[Self.statsKey(account.id, name)],
                   existing.analyzedAt > staleBefore { continue }
                await analyzeNow(account: account, bucket: name)
            }
        }
    }

    func autoAnalyzeAll() {
        guard let account = selectedAccount else { return }
        autoAnalyze(account: account)
    }

    private func scheduleStatsRefresh(bucket: String) {
        // Uploads only carry the bucket name; resolve the owning account,
        // preferring the selected one when names collide across accounts.
        var owner: Account?
        if let selected = selectedAccount,
           bucketList(for: selected.id).contains(where: { $0.name == bucket }) {
            owner = selected
        } else {
            owner = accountStore.accounts.first { account in
                bucketList(for: account.id).contains { $0.name == bucket }
            }
        }
        guard let owner else { return }
        statsRefreshTasks[bucket]?.cancel()
        statsRefreshTasks[bucket] = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 4_000_000_000)
            guard !Task.isCancelled else { return }
            await self?.analyzeNow(account: owner, bucket: bucket)
        }
    }

    static func computeStats(bucket: String, objects: [RemoteObject]) -> BucketStats {
        let files = objects.filter { !$0.isFolder }
        var byExt: [String: (count: Int, size: Int64)] = [:]
        for file in files {
            let ext = file.fileExtension.isEmpty ? "(none)" : file.fileExtension
            var entry = byExt[ext] ?? (0, 0)
            entry.count += 1
            entry.size += file.size
            byExt[ext] = entry
        }
        let extStats = byExt
            .map { ExtensionStat(ext: $0.key, count: $0.value.count, totalSize: $0.value.size) }
            .sorted { $0.totalSize > $1.totalSize }
        return BucketStats(
            bucket: bucket,
            objectCount: files.count,
            totalSize: files.reduce(0) { $0 + $1.size },
            byExtension: extStats,
            largestObjects: Array(files.sorted { $0.size > $1.size }.prefix(10)),
            newestModified: files.compactMap(\.lastModified).max(),
            analyzedAt: Date()
        )
    }
}

File: windows/src/Buckett/ViewModels/AppState.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Avalonia.Threading;
using Buckett.Models;
using Buckett.Services;
using CommunityToolkit.Mvvm.ComponentModel;

namespace Buckett.ViewModels;

/// A bucket qualified by the account it lives in — sidebar sheets need both.
public sealed record AccountBucket(Account Account, Bucket Bucket)
{
    public string Id => Account.Id + "|" + Bucket.Name;
}

public sealed class SidebarSelection : IEquatable<SidebarSelection>
{
    private SidebarSelection(string? bucket) => Bucket = bucket;

    public static readonly SidebarSelection Dashboard = new(null);
    public static SidebarSelection ForBucket(string name) => new(name);

    public string? Bucket { get; }
    public bool IsDashboard => Bucket == null;

    public bool Equals(SidebarSelection? other) => other != null && other.Bucket == Bucket;
    public override bool Equals(object? obj) => Equals(obj as SidebarSelection);
    public override int GetHashCode() => Bucket?.GetHashCode() ?? 0;
}

public sealed class AppState : ObservableObject
{
    public static AppState Shared { get; } = new();

    public AccountStore AccountStore { get; } = new();
    public TransferManager Transfers { get; } = new();
    public UpdateChecker Updates { get; } = new();

    /// Bucket lists for every account (not just the selected one), so the drop
    /// target's menu can offer buckets across accounts.
    private readonly Dictionary<Guid, List<Bucket>> _accountBuckets = new();

    /// Bucket analytics keyed by "accountUUID|bucket" so every account's
    /// buckets can carry stats simultaneously.
    private readonly Dictionary<string, BucketStats> _stats = new();
    private readonly HashSet<string> _analyzing = new();
    private readonly Dictionary<Guid, S3Client> _clientCache = new();
    private readonly Dictionary<string, CancellationTokenSource> _statsRefreshTasks = new();

    private Guid? _selectedAccountID;
    private SidebarSelection _sidebarSelection = SidebarSelection.Dashboard;
    private bool _bucketsLoading;
    private string? _bucketsError;

    /// Raised when bucket lists or stats change, so views can refresh derived lists.
    public event Action? BucketsChanged;
    public event Action? StatsChanged;

    private AppState()
    {
        Transfers.TransferCompleted += bucket => ScheduleStatsRefresh(bucket);

        AccountStore.Accounts.CollectionChanged += (_, _) =>
        {
            OnPropertyChanged(nameof(HasAccounts));
            OnPropertyChanged(nameof(AccountCount));
            BucketsChanged?.Invoke();
        };

        var saved = Settings.Shared.SelectedAccountID;
        if (saved != null && Guid.TryParse(saved, out var id) &&
            AccountStore.Accounts.Any(account => account.Id == id))
        {
            _selectedAccountID = id;
        }
        else
        {
            _selectedAccountID = AccountStore.Accounts.FirstOrDefault()?.Id;
        }
    }

    // MARK: - Selection

    public Guid? SelectedAccountID
    {
        get => _selectedAccountID;
        private set
        {
            if (!SetProperty(ref _selectedAccountID, value)) return;
            Settings.Shared.SelectedAccountID = value?.ToString("D");
            OnPropertyChanged(nameof(SelectedAccount));
        }
    }

    public Account? SelectedAccount =>
        AccountStore.Accounts.FirstOrDefault(account => account.Id == SelectedAccountID);

    public SidebarSelection SidebarSelection
    {
        get => _sidebarSelection;
        set
        {
            if (_sidebarSelection.Equals(value)) return;
            _sidebarSelection = value;
            OnPropertyChanged();
        }
    }

    public bool ShowAllAccounts
    {
        get => Settings.Shared.SidebarShowAllAccounts;
        set
        {
            if (Settings.Shared.SidebarShowAllAccounts == value) return;
            Settings.Shared.SidebarShowAllAccounts = value;
            OnPropertyChanged();
            BucketsChanged?.Invoke();
        }
    }

    public bool HasAccounts => AccountStore.Accounts.Count > 0;
    public int AccountCount => AccountStore.Accounts.Count;

    public bool BucketsLoading
    {
        get => _bucketsLoading;
        private set => SetProperty(ref _bucketsLoading, value);
    }

    public string? BucketsError
    {
        get => _bucketsError;
        private set
        {
            if (SetProperty(ref _bucketsError, value)) OnPropertyChanged(nameof(HasBucketsError));
        }
    }

    public bool HasBucketsError => !string.IsNullOrEmpty(BucketsError);

    // MARK: - Stats

    public static string StatsKey(Guid accountID, string bucket) =>
        accountID.ToString("D").ToUpperInvariant() + "|" + bucket;

    public BucketStats? Stats(Guid accountID, string bucket) =>
        _stats.TryGetValue(StatsKey(accountID, bucket), out var stats) ? stats : null;

    public bool IsAnalyzing(Guid accountID, string bucket) =>
        _analyzing.Contains(StatsKey(accountID, bucket));

    // MARK: - Clients

    public S3Client? Client(Account account)
    {
        if (_clientCache.TryGetValue(account.Id, out var cached)) return cached;
        var client = AccountStore.Client(account);
        if (client != null) _clientCache[account.Id] = client;
        return client;
    }

    public S3Client? CurrentClient => SelectedAccount is { } account ? Client(account) : null;

    // MARK: - Account lifecycle (invalidates cached clients)

    public void SaveAccount(Account account, string? secretKey)
    {
        AccountStore.Upsert(account, secretKey);
        _clientCache.Remove(account.Id);
        if (SelectedAccountID == null) SelectedAccountID = account.Id;
        if (SelectedAccountID == account.Id) _ = LoadBucketsAsync();
        BucketsChanged?.Invoke();
    }

    public void DeleteAccount(Account account)
    {
        AccountStore.Remove(account);
        _clientCache.Remove(account.Id);
        _accountBuckets.Remove(account.Id);
        if (SelectedAccountID == account.Id)
        {
            SelectedAccountID = AccountStore.Accounts.FirstOrDefault()?.Id;
            _stats.Clear();
            SidebarSelection = SidebarSelection.Dashboard;
            _ = LoadBucketsAsync();
        }
        BucketsChanged?.Invoke();
    }

    public void SelectAccount(Guid? id)
    {
        if (id == SelectedAccountID) return;
        SelectedAccountID = id;
        BucketsError = null;
        SidebarSelection = SidebarSelection.Dashboard;
        BucketsChanged?.Invoke();
        _ = LoadBucketsAsync();
    }

    // MARK: - Buckets

    public IReadOnlyList<Bucket> BucketList(Guid accountID) =>
        _accountBuckets.TryGetValue(accountID, out var list)
            ? list
            : Array.Empty<Bucket>();

    public IReadOnlyList<Bucket> Buckets =>
        SelectedAccountID is { } id ? BucketList(id) : Array.Empty<Bucket>();

    public async Task LoadBucketsAsync()
    {
        var account = SelectedAccount;
        if (account == null)
        {
            BucketsChanged?.Invoke();
            return;
        }

        var client = Client(account);
        if (client == null)
        {
            BucketsError =
                $"No credentials found for “{account.DisplayLabel}”. Add the secret key in Settings.";
            _accountBuckets[account.Id] = new List<Bucket>();
            BucketsChanged?.Invoke();
            return;
        }

        BucketsLoading = true;
        BucketsError = null;
        try
        {
            var buckets = await client.ListBucketsAsync().ConfigureAwait(true);
            _accountBuckets[account.Id] = buckets;
            BucketsChanged?.Invoke();
            AutoAnalyze(account);
        }
        catch (Exception error)
        {
            BucketsError = error.Message;
            _accountBuckets[account.Id] = new List<Bucket>();
            BucketsChanged?.Invoke();
        }
        finally
        {
            BucketsLoading = false;
        }

        LoadOtherAccountBuckets();
    }

    /// Background-refreshes bucket lists for accounts other than the selected
    /// one (drop menu, All Accounts sidebar/overview), then kicks off their
    /// stale-stats analysis.
    private void LoadOtherAccountBuckets()
    {
        foreach (var account in AccountStore.Accounts.Where(a => a.Id != SelectedAccountID).ToList())
        {
            var client = Client(account);
            if (client == null) continue;
            _ = Task.Run(async () =>
            {
                try
                {
                    var list = await client.ListBucketsAsync().ConfigureAwait(true);
                    await Dispatcher.UIThread.InvokeAsync(() =>
                    {
                        _accountBuckets[account.Id] = list;
                        BucketsChanged?.Invoke();
                        AutoAnalyze(account);
                    });
                }
                catch (Exception error)
                {
                    Log.Warn($"could not list buckets for {account.DisplayLabel}: {error.Message}");
                }
            });
        }
    }

    public async Task CreateBucketAsync(string name)
    {
        var client = CurrentClient;
        if (client == null) return;
        await client.CreateBucketAsync(name).ConfigureAwait(true);
        await LoadBucketsAsync().ConfigureAwait(true);
    }

    /// Deletes a bucket, optionally emptying it first. Emptying purges ALL
    /// object versions and delete markers (Backblaze B2 keeps hidden versions
    /// that make a bucket look empty while still blocking deletion) and aborts
    /// unfinished multipart uploads. Cleans up aliases, stats, and drop-target
    /// references afterwards.
    public async Task DeleteBucketAsync(string name, Account account, bool emptyFirst)
    {
        var client = Client(account)
            ?? throw new S3Exception(0, "NoCredentials", "Missing account credentials");

        if (emptyFirst)
        {
            var versions = await client.ListAllObjectVersionsAsync(name).ConfigureAwait(true);
            if (versions.Count > 0)
            {
                await client.DeleteObjectVersionsAsync(name, versions).ConfigureAwait(true);
            }
            try
            {
                var uploads = await client.ListMultipartUploadsAsync(name).ConfigureAwait(true);
                foreach (var upload in uploads)
                {
                    try
                    {
                        await client
                            .AbortMultipartUploadAsync(name, upload.Key, upload.UploadID)
                            .ConfigureAwait(true);
                    }
                    catch (Exception error)
                    {
                        Log.Warn($"could not abort upload {upload.UploadID}: {error.Message}");
                    }
                }
            }
            catch (Exception error)
            {
                Log.Warn($"could not list multipart uploads: {error.Message}");
            }
        }

        await client.DeleteBucketAsync(name).ConfigureAwait(true);

        _stats.Remove(StatsKey(account.Id, name));
        BucketAliases.Shared.SetAlias(null, account.Id, name);

        if (_accountBuckets.TryGetValue(account.Id, out var list))
        {
            list.RemoveAll(bucket => bucket.Name == name);
        }
        BucketsChanged?.Invoke();

        if (account.Id == SelectedAccountID)
        {
            if (SidebarSelection.Bucket == name) SidebarSelection = SidebarSelection.Dashboard;
            await LoadBucketsAsync().ConfigureAwait(true);
        }
    }

    /// Raised when something outside the main window needs it brought forward.
    public event Action? OpenMainWindowRequested;

    public void OpenMainWindow() => OpenMainWindowRequested?.Invoke();

    // MARK: - Analytics

    /// Convenience for the selected account.
    public void Analyze(string bucket)
    {
        if (SelectedAccount is { } account) Analyze(account, bucket);
    }

    public void Analyze(Account account, string bucket) => _ = AnalyzeNowAsync(account, bucket);

    public async Task AnalyzeNowAsync(Account account, string bucket)
    {
        var key = StatsKey(account.Id, bucket);
        var client = Client(account);
        if (client == null || _analyzing.Contains(key)) return;

        _analyzing.Add(key);
        StatsChanged?.Invoke();
        try
        {
            var objects = await client.ListAllObjectsAsync(bucket).ConfigureAwait(true);
            _stats[key] = ComputeStats(bucket, objects);
        }
        catch (Exception error)
        {
            Log.Warn($"analyze failed for {bucket}: {error.Message}");
        }
        finally
        {
            _analyzing.Remove(key);
            StatsChanged?.Invoke();
        }
    }

    /// Analyzes the account's buckets that have no stats yet or whose stats
    /// are older than 15 minutes. Sequential so accounts aren't hammered.
    public void AutoAnalyze(Account account)
    {
        var staleBefore = DateTime.UtcNow.AddMinutes(-15);
        var names = BucketList(account.Id).Select(bucket => bucket.Name).ToList();
        _ = AutoAnalyzeSequentiallyAsync(account, names, staleBefore);
    }

    /// One bucket at a time — a wide account shouldn't fire dozens of full
    /// listings at once.
    private async Task AutoAnalyzeSequentiallyAsync(
        Account account, IReadOnlyList<string> names, DateTime staleBefore)
    {
        foreach (var name in names)
        {
            if (_stats.TryGetValue(StatsKey(account.Id, name), out var existing) &&
                existing.AnalyzedAt > staleBefore)
            {
                continue;
            }
            await AnalyzeNowAsync(account, name).ConfigureAwait(true);
        }
    }

    public void AutoAnalyzeAll()
    {
        if (SelectedAccount is { } account) AutoAnalyze(account);
    }

    private void ScheduleStatsRefresh(string bucket)
    {
        // Uploads only carry the bucket name; resolve the owning account,
        // preferring the selected one when names collide across accounts.
        Account? owner = null;
        if (SelectedAccount is { } selected &&
            BucketList(selected.Id).Any(candidate => candidate.Name == bucket))
        {
            owner = selected;
        }
        else
        {
            owner = AccountStore.Accounts.FirstOrDefault(account =>
                BucketList(account.Id).Any(candidate => candidate.Name == bucket));
        }
        if (owner == null) return;

        if (_statsRefreshTasks.TryGetValue(bucket, out var existing))
        {
            existing.Cancel();
            existing.Dispose();
        }
        var cancellation = new CancellationTokenSource();
        _statsRefreshTasks[bucket] = cancellation;

        _ = Task.Run(async () =>
        {
            try
            {
                await Task.Delay(TimeSpan.FromSeconds(4), cancellation.Token).ConfigureAwait(true);
                await Dispatcher.UIThread
                    .InvokeAsync(async () => await AnalyzeNowAsync(owner, bucket))
                    .ConfigureAwait(true);
            }
            catch (OperationCanceledException)
            {
                // Superseded by a newer upload into the same bucket.
            }
        });
    }

    public static BucketStats ComputeStats(string bucket, IReadOnlyList<RemoteObject> objects)
    {
        var files = objects.Where(item => !item.IsFolder).ToList();
        var byExtension = files
            .GroupBy(file => file.FileExtension.Length == 0 ? "(none)" : file.FileExtension)
            .Select(group => new ExtensionStat(
                group.Key, group.Count(), group.Sum(file => file.Size)))
            .OrderByDescending(stat => stat.TotalSize)
            .ToList();

        return new BucketStats
        {
            Bucket = bucket,
            ObjectCount = files.Count,
            TotalSize = files.Sum(file => file.Size),
            ByExtension = byExtension,
            LargestObjects = files.OrderByDescending(file => file.Size).Take(10).ToList(),
            NewestModified = files
                .Where(file => file.LastModified.HasValue)
                .Select(file => file.LastModified!.Value)
                .DefaultIfEmpty()
                .Max() is var newest && newest == default ? null : newest,
            AnalyzedAt = DateTime.UtcNow
        };
    }
}
